bigframes/functions/remote_function.py

```diff
@@ -285,38 +289,104 @@ def generate_cloud_function_main_code(self, def_, dir):
-              replies.append(reply)
-            return_json = json.dumps({{"replies" : replies}})
-            return return_json
-          except Exception as e:
-            return jsonify( {{ "errorMessage": str(e) }} ), 400
-        """
-        )
-
-        code = code_template.format(
-            udf_code_file=udf_code_file,
-            udf_bytecode_file=udf_bytecode_file,
-            handler_func_name=handler_func_name,
-        )
+        code = """\
+import cloudpickle
+import functions_framework
+from flask import jsonify
+import json
+"""
+        if is_row_processor:
+            code += """\
+import pandas as pd
+
+def get_pd_series(row):
+    row_json = json.loads(row)
+    col_names = row_json["names"]
+    col_types = row_json["types"]
+    col_values = row_json["values"]
+    index_names = row_json["index"]
+    dtype = row_json["dtype"]
+
+    # index and column names are not necessarily strings
+    # they are serialized as repr(repr(name)) at source
+    col_names = [eval(name) for name in col_names]
+    index_names = [eval(name) for name in index_names]
+
+    value_converters = {
+        "boolean": lambda val: val == "true",
+        "Int64": int,
+        "Float64": float,
+        "string": str,
+    }
+
+    def convert_value(value, value_type):
+        value_converter = value_converters.get(value_type)
+        if value_converter is None:
+            raise ValueError(f"Don't know how to handle type '{value_type}'")
+        if value is None:
+            return None
+        return value_converter(value)
+
+    index_positions = [col_names.index(col) for col in index_names]
+
+    index_values = [
+        pd.Series([convert_value(col_values[i], col_types[i])], dtype=col_types[i])[0]
+        for i in index_positions
+    ]
+
+    col_names = [
+        col_names[i]
+        for i, a in enumerate(col_names)
+        if i not in index_positions
+    ]
+
+    row_values = [
+        pd.Series([convert_value(a, col_types[i])], dtype=col_types[i])[0]
+        for i, a in enumerate(col_values)
+        if i not in index_positions
+    ]
+
+    row_index = index_values[0] if len(index_values) == 1 else tuple(index_values)
+    row_series = pd.Series(row_values, index=col_names, name=row_index, dtype=dtype)
+    return row_series
+"""
+        code += f"""\
+
+# original udf code is in {udf_code_file}
+# serialized udf code is in {udf_bytecode_file}
+with open("{udf_bytecode_file}", "rb") as f:
+    udf = cloudpickle.load(f)
+
+def {handler_func_name}(request):
+    try:
+        request_json = request.get_json(silent=True)
+        calls = request_json["calls"]
+        replies = []
+        for call in calls:
+"""
+        if is_row_processor:
+            code += """\
+            reply = udf(get_pd_series(call[0]))
+            if pd.isna(reply):
+                # Pandas N/A values are not json serializable, so use a python
+                # equivalent instead
+                reply = None
```

Comment:
Do we need to differentiate NaN vs None for float dtype?

Reply:
It should be taken care by BigFrames -> BQ contract, where they both are represented the same (NULL) in BQ, and in cloud function they are deserialized as python `None`.

```python
import pandas as pd
import numpy as np

df = pd.Series([1, 2.5, pd.NA, np.nan, None])
df
```
```
0       1
1     2.5
2    <NA>
3     NaN
4    None
dtype: object
```
```python
import bigframes.pandas as bpd
bpd.options.bigquery.project = "bigframes-dev"

bdf = bpd.read_pandas(df)
bdf
```
```
0     1.0
1     2.5
2    <NA>
3    <NA>
4    <NA>
dtype: Float64
```

Reply:
That is a limitation of our type-widening logic as we go from non-nullable float to nullable float. In this case here, we are mapping nullable float to nullable float. In nullable float type, nan is not the null value, so I think we should be able to keep it separate.

Reply:
handled (nan, inf, -inf) as special values, PTAL